torch/distributed/launch.py

```diff
@@ -221,7 +248,7 @@ def main():
     # set PyTorch distributed related environmental variables
     current_env = os.environ.copy()
     current_env["MASTER_ADDR"] = args.master_addr
-    current_env["MASTER_PORT"] = str(args.master_port)
+    current_env["MASTER_PORT"] = str(get_port(args.master_addr, args.master_port))
```

Comment:
I don't think we need `get_port`, just setting "MASTER_PORT" to `args.master_port` should suffice. When "MASTER_PORT" is 0, it would pass this value all the way down to TCPStore: https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/Utils.cpp#L80, which would listen on port 0 which would automatically pick a free port.

This also gets rid of the race as well. I'm wondering if we need any code change at all here since `--master_port 0` should work appropriately? We probably just need to update the docs to reflect this?

Reply:
Oh, I had no idea it was that simple - then yes, by all means, this is a much more superior solution. 

I wonder why it wasn't used in first place, in used a hardcoded 29500 port instead.

Reply:
probably should also document that 29500 is the default port - I don't think it says that anywhere in the public docs.

Reply:
Actually there is a problem with passing in 0 as master port that I missed while writing the original comment. This would only work for a single process. Passing in a port of 0 means that every process will probably end up choosing a different port and that would cause rendezvous to fail since all processes need to agree on a single master port and address that they need to connect to.

I couldn't think of a nice way to enable dynamic ports here since the static master port and address is essentially how we ensure all processes can start communicating in the first place.

Reply:
Oh bummer, I thought it was some magical way, but as you explain it - it's not.

To solve the race solution there needs to be some sort of port server that can allocate and hold the port for the client to connect later.

The impetus for this PR was a need for more than one test in the test suite being able to launch ddp.

Reply:
> The impetus for this PR was a need for more than one test in the test suite being able to launch ddp.

I see, in that case I guess the test_suite itself should probably perform some logic similar to `get_port` in this PR and pass an appropriate MASTER_PORT to the launcher.

Reply:
Totally, but I thought why re-invent the wheel thousands of times for each user when the framework could provide this solution.

That's why my 2nd proposal in OP, is to provide a flag which will randomize and check the port. this won't avoid the race condition though so I wasn't sure this is good enough still. And having an unreliable test suite isn't good. So the test suite needs to solve it on the client side with some internal counter then.

So I'm totally fine if this PR is put to rest as an experiment.